### DISCUSSION TITLE
Client disconnect() blocks main thread for a long time

### DISCUSSION THREAD

**Discussion by jsib0** [miguelgrinberg/python-socketio#739] [Q&A]
Hello.

I am trying to disconnect the socket connection when I detect that the WiFi of my machine is disconnected. However, when I perform a disconnect() when the WiFi is disconnected, I receive the following logs below then the main thread stops right after. It takes time for the main thread to revive... Any directions to move forward is highly appreciated. Thank you in advance.

**Client logs:**

```
Sending packet MESSAGE data 1/socket
Sending packet MESSAGE data 1
Sending packet CLOSE data None
Exiting Write loop task
Engine.IO Connection dropped
Exiting ping task
```

**Code to disconnect**

```
    def attempt_connection(self):
        logger.info("Attempting connection")
        if self.net.has_internet:
            if self.client.connected:
                return True
            try:
                self.client.connect()
            except ConnectionError:
                return False
        elif self.client.connected:
            try:
                self.client.disconnect()
            except:
                logger.info("except")

        return False

## client class
Class Client:
    def disconnect(self):
        self.sio_client.disconnect()
        self.sio_client.wait()
        return True
```

**Comment by miguelgrinberg** [comment]
The `disconnect` function that you are using is supposed to be used for a graceful disconnect. This involves sending a `CLOSE` packet to the other side. Since you are doing this when you don't have internet, this isn't going to work. It would be best to leave the connection alone, and eventually the client itself will realize it's been disconnected.

**Comment by jsib0** [comment]
@miguelgrinberg Thank you for the quick response!  Is there any way at all to control the disconnect on the developer's end?

**Comment by miguelgrinberg** [comment]
@jsib0 I don't understand what you mean by "control the disconnect". You are trying to disconnect from the server when there is no connectivity, that's the problem. At that point you are already disconnected. Are you using a disconnect handler? The handler will be called when the disconnection is detected.

**Comment by jsib0** [comment]
What I meant was to clean up the processes using of socket Client. Since calling `disconnect` blocks and stalls the main thread, I wanted to see if there is another way to cleanly disconnect without having to use the `disconnect` method

**Comment by miguelgrinberg** [comment]
@jsib0 You did not answer my question. Do you have a disconnect handler set up? Is it not getting called?

**Comment by jsib0** [comment]
When I call `disconnect()` the disconnect handler on the client side does not run. When I remove the `disconnect()` call, but place a `reconnection_attempts` of 1, then when the internet is cut, the `on_disconnect` handler is called.

```
def on_disconnect(self):
    self.sio_client.logger.info(f"Disconnecting...")
    return True
```

**Comment by miguelgrinberg** [comment]
Calling disconnect when you don't have connectivity is not really a use case that I envisioned, so let's set that one aside and look at the other use cases.

Are you implying that the disconnect handler is only called when you have a single reconnection attempt? What happens when you have more than one?

**Comment by jsib0** [comment]
I noted that to imply that if I used this path as a way to disconnect, that it works the way I would like to. And it does call the disconnect handler. I am sure the disconnect handler will also be triggered if I adjusted it to 10. I tried 3 and it works.
I tested with the `_handle_eio_disconnect` and it seems to work how I would like it to work. Do you have any concerns with using this function directly?

**Comment by miguelgrinberg** [comment]
@jsib0 I still don't understand what's broken. As I said, the client handles abrupt disconnections on its own, it does not need help from you. Anything else you try to do will interfere.

You are now mentioning a lot of use cases that do work as expected. Can you please tell me what are the use cases that do not work, so that I determine if there is a bug or user error?

Also, you shouldn't be tapping into the low level Engine.IO handlers. Those are not a public part of the API and I reserve the right to change them at any time without deprecation or actually any notice.

**Comment by jsib0** [comment]
@miguelgrinberg I see what you mean. What I am trying to do is relay to the user whether the device is connected to the server, which is through the socket connection. On the device, the user is able to see the wifi connection as well. If the wifi is disconnected, a server connection is technically not possible, thus not true; however, python-socketio maintains a "connected" state until the amount of reconnection tries is exhausted. With all the python-socketio and flask-socketio settings, the connected state would stay for about 25 seconds.

There are work-arounds in this problem, but we wanted a true source of truth to be relayed to the user - this also gives us the assurance that the device is at a true disconnected state.

**Comment by miguelgrinberg** [comment]
> however, python-socketio maintains a "connected" state until the amount of reconnection tries is exhausted.

This is actually incorrect. The connect and disconnect handlers are called for each connection. It does not matter that you have retries enabled, when a connection is gone, the disconnect handler is called, then the retries begin, and once a connection is reestablished you will get another call on the connect handler.

It is really up to the networking stack on your device when a connection is reported as gone. In general this only happens when you read or write from the socket, there is no actual disconnect notification in low-level networking. The 25 second delay is because you probably aren't sending any data on this socket, so the disconnection is detected when the ping/pong exchange is attempted, every 25 seconds.

If you find out about the networking stack going away through a different mechanism, then the best thing you can do is let Socket.IO do its thing, as it will eventually recognize that the network is gone.
